### PR TITLE
refactor: use node:fs/promises

### PR DESCRIPTION
**Pull request by alumni**
### Description of change

Replacing the sync functions from `node:fs` with async functions from `node:fs/promises`. This should simplify the code by not requiring Promise wrappers and it should not block the event loop anymore.

### Pull-Request Checklist

- [X] Code is up-to-date with the `master` branch
- [X] `npm run format` to apply prettier formatting
- [X] `npm run test` passes with this change
- [ ] This pull request links relevant issues as `Fixes #0000`
- [X] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [X] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: test/functional/repository/find-options/repository-find-options.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
-import "reflect-metadata"
```
Comment (nikelborm) [inline test/functional/repository/find-options/repository-find-options.ts:1]: Why did you remove reflect metadata here? Is it really not necessary?

File: test/functional/repository/find-options/repository-find-options.ts [hunk 2]
```diff
@@ -256,9 +253,9 @@ describe("repository > find options > comment", () => {
     beforeEach(() => reloadTestingDatabases(connections))
     after(async () => {
         await closeTestingConnections(connections)
-        if (fs.existsSync(logPath)) {
-            await promisify(unlink)(logPath)
-        }
+        try {
+            fs.unlink(logPath)
```
Comment (jorenvandeweyer) [inline test/functional/repository/find-options/repository-find-options.ts:257]: await

File: src/commands/CommandUtils.ts [hunk 3]
```diff
@@ -79,26 +78,26 @@ export class CommandUtils {
...
     /**
      * Reads everything from a given file and returns its content as a string.
      */
     static async readFile(filePath: string): Promise<string> {
-        return new Promise<string>((ok, fail) => {
-            fs.readFile(filePath, (err, data) =>
-                err ? fail(err) : ok(data.toString()),
-            )
-        })
+        return (await fs.readFile(filePath)).toString()
```
Comment (jorenvandeweyer) [inline src/commands/CommandUtils.ts:91]: Split this into 2 lines

File: src/commands/CommandUtils.ts [hunk 4]
```diff
@@ -67,7 +66,7 @@ export class CommandUtils {
      * Creates directories recursively.
      */
     static createDirectories(directory: string) {
```
Comment (jorenvandeweyer) [inline src/commands/CommandUtils.ts:68]: Updaye return type to Promise

File: src/util/ImportUtils.ts [hunk 5]
```diff
@@ -39,45 +39,31 @@ export async function importOrRequireFile(
...
-                                        accept(null)
-                                    }
-                                }
-                            },
-                        )
-                    } else searchPackageJson()
-                })
+        try {
+            const stats = await fs.stat(potentialPackageJson)
+            if (stats.isFile()) {
```
Comment (jorenvandeweyer) [inline src/util/ImportUtils.ts:51]: Use an early return here for better readability.

```
if (!stats.isFile()) {
  continue
}
```

File: src/util/ImportUtils.ts [hunk 6]
```diff
@@ -39,45 +39,31 @@ export async function importOrRequireFile(
...
+                    return JSON.parse(
+                        await fs.readFile(potentialPackageJson, "utf8"),
+                    )
+                } catch {
+                    return null
+                }
+            } else {
+                continue
             }
+        } catch {
```
Comment (jorenvandeweyer) [inline src/util/ImportUtils.ts:62]: What are you catching here? A potential throw from fs.stats?

File: test/functional/util/ImportUtils.ts [hunk 7]
```diff
@@ -41,7 +38,7 @@ describe("ImportUtils.importOrRequireFile", () => {
         expect(exports.default).to.be.a("function")
         expect(exports.number).to.be.eq(6)
 
-        rmdirSync(testDir)
+        await fs.rmdir(testDir, { recursive: true })
```
Comment (jorenvandeweyer) [inline test/functional/util/ImportUtils.ts:41]: Indentation

**Review by jorenvandeweyer — CHANGES_REQUESTED** [review]

File: test/functional/repository/find-options/repository-find-options.ts (hunk 1 above)
Comment (alumni) [reply test/functional/repository/find-options/repository-find-options.ts:1]: Not necessary. `reflect-metadata` only needs importing once.

File: src/commands/CommandUtils.ts (hunk 3 above)
Comment (alumni) [reply src/commands/CommandUtils.ts:91]: split how into two lines?

File: src/util/ImportUtils.ts (hunk 6 above)
Comment (alumni) [reply src/util/ImportUtils.ts:62]: The same thing that was caught before: https://nodejs.org/api/errors.html#common-system-errors

File: test/functional/util/ImportUtils.ts (hunk 7 above)
Comment (alumni) [reply test/functional/util/ImportUtils.ts:41]: Indentation is correct. The file was formatted before commit.

File: test/functional/repository/find-options/repository-find-options.ts (hunk 2 above)
Comment (alumni) [reply test/functional/repository/find-options/repository-find-options.ts]: This was a good catch, I somehow didn't see it. 

There's an ESLint rule for this: `@typescript-eslint/no-floating-promises`, and it's in the `recommended-type-checked` config. We should consider enabling these rules.

Comment (jorenvandeweyer) [reply test/functional/repository/find-options/repository-find-options.ts]: absolutely

File: src/commands/CommandUtils.ts (hunk 3 above)
Comment (jorenvandeweyer) [reply src/commands/CommandUtils.ts:91]:
```typescript
const file = await fs.readFile(filePath))

return file.toString()
```

**Review by jorenvandeweyer — APPROVED** [review]

File: test/functional/repository/find-options/repository-find-options.ts (hunk 1 above)
Comment (alumni) [reply test/functional/repository/find-options/repository-find-options.ts:1]: If `test/utils/test-setup.ts` is loaded correctly by mocha before the tests, this should definitely not be needed. I noticed some files also load `test-setup.ts` by themselves, this should also not be required.

It looks like a bigger cleanup is required for the test setup, so I reverted this for now.

**Review by michaelbromley — APPROVED** [review]
